File: 40.cpp
class Solution {
private:
    vector<vector<int>> res;
    vector<int> vec;
    int tar;

public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        tar = target;
        helper(candidates, 0, 0);
        return res;
    }
    
    void helper(vector<int>&can, int pos, int sum){
        for(int i=pos; i<can.size(); i++){
            if(i>pos && can[i-1]==can[i]) 
                continue;  //过滤掉重复的方案
            int s = sum + can[i];
            if(s==tar){
                vec.push_back(can[i]);
                res.push_back(vec);
                vec.pop_back();
                return;
            }else if(s>tar){
                return;
            }else{
                vec.push_back(can[i]);
                helper(can, i+1, s);
                vec.pop_back();
            }
        }
    }
};

File: golang/40/40.go
package main

import "sort"

var res [][]int
var cur []int

func combinationSum2(candidates []int, target int) [][]int {
	sort.Ints(candidates)

	res = make([][]int, 0)
	cur = make([]int, 0)

	backtrack(candidates, 0, 0, target)

	return res
}

func backtrack(nums []int, start, sum, target int) {
	if sum == target {
		tmp := make([]int, len(cur))
		copy(tmp, cur)
		res = append(res, tmp)
		return
	}

	if sum > target {
		return
	}

	for i := start; i < len(nums); i++ {
		if i > start && nums[i] == nums[i-1] {
			continue
		}
		cur = append(cur, nums[i])
		backtrack(nums, i+1, sum+nums[i], target)
		cur = cur[:len(cur)-1]
	}
}
